presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1294,12 +1302,17 @@ private Partitioning createPartitioning(List<Symbol> partitioningColumns)
             return Partitioning.create(partitioningHandle, partitioningColumns);
         }
 
-        private Scope selectExchangeScopeForPartitionedRemoteExchange(PlanNode exchangeSource)
+        // TODO: refactor this method into ExchangeNode#partitionedExchange once
+        //   materialized exchange is supported for all nodes.
+        private Scope selectExchangeScopeForPartitionedRemoteExchange(PlanNode exchangeSource, boolean nullsAndAnyReplicated)
         {
-            if (exchangeSource.getOutputSymbols().isEmpty()) {
-                // materializing 0 columns input is not supported
+            if (nullsAndAnyReplicated || exchangeSource.getOutputSymbols().isEmpty()) {
+                // materialized remote exchange is not supported when
+                //  * replicateNullsAndAny is needed
```

Comment:
For those like me who have always been confused by replicateNullsAndAny, here's more information about it taken from https://github.com/prestodb/presto/pull/7792/commits/72fb3611124a92400f2d9d04dafc925de2991ccf#r116876456

>We check the following conditions in semi join:
>* if there is matching value (for this regular partitioning & redistribution works)
>* else, if there are nulls on build side (for this replication of nulls works)
>* else, if [probe is null and] there is anything on build side (for this replicating anything, regardless of its nullity, works)

So basically in order to have correct results for a semi join we can't just repartition
1.  if there are **nulls** on the build side those need to be replicated to every node so that all partitions produce the correct output 
2. at least one row (**any** row) from the build side needs to be replicated to all the worker nodes in case there are nulls on the probe side. 

Hence `replicateNullsAndAny`

Reply:
@rschlussel : Thanks for the explanation ! Consider adding comment into `PartitioningScheme` ?